Auth tokens configured by `pnpm config set` are not found by npm

The version `7.28.0` doesn't work, but the `7.27.1` works. Here's the relevant part of the config to make it work:

```yaml
publish:
  stage: publish
  before_script:
    - corepack enable
    - corepack prepare pnpm@7.27.1 --activate
    - pnpm config set "//${GITLAB_REGISTRY}:_authToken=${CI_JOB_TOKEN}"
```

_Originally posted by @navaru in https://github.com/pnpm/pnpm/issues/4937#issuecomment-1455139283_

This happens because `pnpm config set` writes to its own rc file at `~/.config/pnpm/rc`, which is not read by npm CLI (and pnpm publish run `npm publish` under the hood).

I think the solution should be to make pnpm run `npm config set` when auth tokens are configured.

gitlab-ci.yml

```yaml
image: node:latest

stages:
  - deploy

deploy:
  stage: deploy
  before_script:
    - corepack enable
    - corepack prepare pnpm@latest --activate
    - npm config set "//${CI_SERVER_HOST}/api/v4/projects/${CI_PROJECT_ID}/packages/npm/:_authToken=${CI_JOB_TOKEN}"
  script:
    - pnpm install --frozen-lockfile --strict-peer-dependencies
    - pnpm publish --publish-branch main --no-git-checks
```

I thought I'd double-check this doesn't work. No luck, I'm afraid.

```bash
npm ERR! code ENEEDAUTH
npm ERR! need auth This command requires you to be logged in to https://gitlab.com/api/v4/projects/43729055/packages/npm/
npm ERR! need auth You need to authorize this machine using `npm adduser`
npm ERR! A complete log of this run can be found in:
npm ERR!     /root/.npm/_logs/2023-03-08T00_50_37_922Z-debug-0.log
```

--- a/.changeset/nervous-dots-sleep.md
+++ b/.changeset/nervous-dots-sleep.md
@@ -1,0 +1,6 @@
+---
+"@pnpm/plugin-commands-config": patch
+"pnpm": patch
+---
+
+Settings related to authorization should be set/deleted by npm CLI [#6181](https://github.com/pnpm/pnpm/issues/6181).

--- a/config/plugin-commands-config/package.json
+++ b/config/plugin-commands-config/package.json
@@ -33,6 +33,7 @@
     "@pnpm/cli-utils": "workspace:*",
     "@pnpm/config": "workspace:*",
     "@pnpm/error": "workspace:*",
+    "@pnpm/run-npm": "workspace:*",
     "ini": "3.0.1",
     "read-ini-file": "4.0.0",
     "render-help": "^1.0.3",

--- a/config/plugin-commands-config/src/ConfigCommandOptions.ts
+++ b/config/plugin-commands-config/src/ConfigCommandOptions.ts
@@ -5,6 +5,7 @@ export type ConfigCommandOptions = Pick<Config,
 | 'cliOptions'
 | 'dir'
 | 'global'
+| 'npmPath'
 | 'rawConfig'
 > & {
   json?: boolean

--- a/config/plugin-commands-config/src/configSet.ts
+++ b/config/plugin-commands-config/src/configSet.ts
@@ -1,10 +1,20 @@
 import path from 'path'
+import { runNpm } from '@pnpm/run-npm'
 import { readIniFile } from 'read-ini-file'
 import { writeIniFile } from 'write-ini-file'
 import { ConfigCommandOptions } from './ConfigCommandOptions'
 
 export async function configSet (opts: ConfigCommandOptions, key: string, value: string | null) {
   const configPath = opts.global ? path.join(opts.configDir, 'rc') : path.join(opts.dir, '.npmrc')
+  if (opts.global && settingShouldFallBackToNpm(key)) {
+    const _runNpm = runNpm.bind(null, opts.npmPath)
+    if (value == null) {
+      _runNpm(['config', 'delete', key])
+    } else {
+      _runNpm(['config', 'set', `${key}=${value}`])
+    }
+    return
+  }
   const settings = await safeReadIniFile(configPath)
   if (value == null) {
     if (settings[key] == null) return
@@ -13,6 +23,14 @@ export async function configSet (opts: ConfigCommandOptions, key: string, value:
     settings[key] = value
   }
   await writeIniFile(configPath, settings)
+}
+
+function settingShouldFallBackToNpm (key: string): boolean {
+  return (
+    ['registry', '_auth', '_authToken', 'username', '_password'].includes(key) ||
+    key[0] === '@' ||
+    key.startsWith('//')
+  )
 }
 
 async function safeReadIniFile (configPath: string): Promise<Record<string, unknown>> {

--- a/config/plugin-commands-config/test/managingAuthSettings.test.ts
+++ b/config/plugin-commands-config/test/managingAuthSettings.test.ts
@@ -1,0 +1,39 @@
+import { config } from '@pnpm/plugin-commands-config'
+import { runNpm } from '@pnpm/run-npm'
+
+jest.mock('@pnpm/run-npm', () => ({
+  runNpm: jest.fn(),
+}))
+
+describe.each(
+  [
+    '_auth',
+    '_authToken',
+    '_password',
+    'username',
+    'registry',
+    '@foo:registry',
+    '//registry.npmjs.org/:_authToken',
+  ]
+)('settings related to auth are handled by npm CLI', (key) => {
+  it(`should set ${key}`, async () => {
+    await config.handler({
+      dir: process.cwd(),
+      cliOptions: {},
+      configDir: __dirname, // this doesn't matter, it won't be used
+      rawConfig: {},
+    }, ['set', `${key}=123`])
+
+    expect(runNpm).toHaveBeenCalledWith(undefined, ['config', 'set', `${key}=123`])
+  })
+  it(`should delete ${key}`, async () => {
+    await config.handler({
+      dir: process.cwd(),
+      cliOptions: {},
+      configDir: __dirname, // this doesn't matter, it won't be used
+      rawConfig: {},
+    }, ['delete', key])
+
+    expect(runNpm).toHaveBeenCalledWith(undefined, ['config', 'delete', key])
+  })
+})

--- a/config/plugin-commands-config/test/managingAuthSettings.test.ts
+++ b/config/plugin-commands-config/test/managingAuthSettings.test.ts
@@ -16,24 +16,18 @@ describe.each(
     '//registry.npmjs.org/:_authToken',
   ]
 )('settings related to auth are handled by npm CLI', (key) => {
+  const configOpts = {
+    dir: process.cwd(),
+    cliOptions: {},
+    configDir: __dirname, // this doesn't matter, it won't be used
+    rawConfig: {},
+  }
   it(`should set ${key}`, async () => {
-    await config.handler({
-      dir: process.cwd(),
-      cliOptions: {},
-      configDir: __dirname, // this doesn't matter, it won't be used
-      rawConfig: {},
-    }, ['set', `${key}=123`])
-
+    await config.handler(configOpts, ['set', `${key}=123`])
     expect(runNpm).toHaveBeenCalledWith(undefined, ['config', 'set', `${key}=123`])
   })
   it(`should delete ${key}`, async () => {
-    await config.handler({
-      dir: process.cwd(),
-      cliOptions: {},
-      configDir: __dirname, // this doesn't matter, it won't be used
-      rawConfig: {},
-    }, ['delete', key])
-
+    await config.handler(configOpts, ['delete', key])
     expect(runNpm).toHaveBeenCalledWith(undefined, ['config', 'delete', key])
   })
 })

--- a/config/plugin-commands-config/tsconfig.json
+++ b/config/plugin-commands-config/tsconfig.json
@@ -16,6 +16,9 @@
       "path": "../../cli/cli-utils"
     },
     {
+      "path": "../../exec/run-npm"
+    },
+    {
       "path": "../../packages/error"
     },
     {

